File: produban-spark-streaming-scala/src/main/scala/com/produban/sparkstreaming/AkamaiKafkaStreaming.scala
package com.produban.sparkstreaming

import scala.collection.JavaConverters._
import org.apache.spark.SparkContext
import org.apache.spark.SparkContext._
import org.apache.spark.SparkConf
import akka.dispatch.Foreach
import org.apache.spark.streaming.StreamingContext
import StreamingContext._
import org.apache.spark.streaming.Seconds
import org.apache.spark.rdd.PairRDDFunctions
import com.produban.util.JsonUtil
import org.apache.spark.streaming.dstream.ReceiverInputDStream
import org.apache.commons.collections.IteratorUtils
import com.produban.indexer.elastic.ElasticIndexer
import com.produban.akamai.api.Alert
import com.produban.akamai.entity.Akamai
import com.produban.api.general.K
import com.produban.util.FilePropertiesUtil
import org.apache.commons.lang.SystemUtils
import java.util.Properties
import com.produban.indexer.api.Indexer
import org.apache.spark.streaming.kafka.KafkaUtils
import com.produban.akamai.api.AlertElastic
import java.util.UUID
import com.produban.api.data.Rule

class AkamaiKafkaStreaming(indexer: Indexer) extends Serializable {

  def executeRules(configuration: Properties) {

    val sparkPort = configuration.getProperty(K.SYSTEM.PROPERTY_SPARKSTREAMING_PORT)
    val sparkMode = configuration.getProperty(K.SYSTEM.PROPERTY_SPARKSTREAMING_MODE)
    val sparkCheckPoint = configuration.getProperty(K.SYSTEM.PROPERTY_SPARKSTREAMING_CHECKPOINT)
    val zookeeperNodes = configuration.getProperty(K.KAFKA.PROPERTY_ZOOKEEPER_NODES)
    val topicsKafka = Map("flume-topic" -> 1)

    //If we want to configure mode and name with code, bad idea... better pass like configuration
    //val sparkConf = new SparkConf().setMaster(sparkMode).setAppName("MiApp");
    val sparkConf = new SparkConf().setMaster(sparkMode);
    val ssc = new StreamingContext(sparkConf, Seconds(5));
    val lines = KafkaUtils.createStream(ssc, zookeeperNodes, "myGroup", topicsKafka)

    var rule = new Rule()
    rule.setMessage("Alert SQL Injection")
    rule.setNumberTimes(3)
    rule.setRegex("SQL")
    rule.setWindow(15)
    rule.setSlideWindow(5)
    val rules = Array(rule)
    
    executeRules(lines, rules)

    // ssc.checkpoint(sparkCheckPoint)
    ssc.start()
    ssc.awaitTermination()

  }

  def executeRules(lines: ReceiverInputDStream[(String, String)], rules: Array[Rule]) {
    for (rule <- rules) {
      val filterSql = lines.map(line => line._2).
        filter(line => line.contains(rule.getRegex())).
        map {
          line =>
            JsonUtil.read(line.getBytes(), classOf[Akamai])
        }

      val groupSql = filterSql.map {
        json =>
          val srcIp = json.getMessage().getCliIP()
          val srcURL = json.getMessage().getReqHost()
          (srcIp + "_" + srcURL, json)
      }

      val errorLinesValueReduce = groupSql.groupByKeyAndWindow(Seconds(rule.getWindow()), Seconds(rule.getSlideWindow()));
      errorLinesValueReduce.foreachRDD {
        rdd =>
          val elem1 = rdd.take(1)

          if (elem1.size > 0) {

            val alertsAkamai = elem1(0)._2
            if (alertsAkamai.size > rule.getNumberTimes()) {
              //Todos los logs en un array de elementos dentro de AlertS   
              val jsonsToIndex = createAlert(rule.getMessage(), alertsAkamai.toList)
              indexer.indexJsons(jsonsToIndex)
            }
          }
      }
    }
  }

  private def createAlert(message: String, jsonsDetected: List[Akamai]): Array[String] = {
    val idUUID = UUID.randomUUID().toString()
    var alertJsons: Array[String] = new Array[String](jsonsDetected.size)
    for (index <- 0 until jsonsDetected.size) {
      val alert: AlertElastic = new AlertElastic(idUUID, message, jsonsDetected(index))
      alertJsons(index) = JsonUtil.write(alert)
    }

    return alertJsons
  }

  private def manageInput(args: Array[String]): Properties = {
    var configuration: Properties = new Properties()
    //    if (args.length < 1) {
    //      println("Error parameter missing, we need to indicate LOCAL||VIRT||PROD")
    //      exit(-1)
    //    } else {
    if (SystemUtils.IS_OS_WINDOWS) {
      configuration = FilePropertiesUtil.getProperties(K.SYSTEM.CONFIG_FILE_WINDOWS)
    } else {
      if (args(0).equals("VIRT")) {
        configuration = FilePropertiesUtil.getProperties(K.SYSTEM.CONFIG_FILE_LINUX_VIRT)
      } else {
        configuration = FilePropertiesUtil.getProperties(K.SYSTEM.CONFIG_FILE_LINUX_PROD)
      }

    }
    //    }
    return configuration
  }

}

File: produban-spark-streaming/src/main/java/com/produban/starkstreaming/AkamaiKafkaStreaming.java
package com.produban.starkstreaming;

import java.io.Serializable;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

import org.apache.commons.collections.IteratorUtils;
import org.apache.spark.SparkConf;
import org.apache.spark.api.java.JavaPairRDD;
import org.apache.spark.api.java.function.Function;
import org.apache.spark.api.java.function.PairFunction;
import org.apache.spark.streaming.Duration;
import org.apache.spark.streaming.api.java.JavaDStream;
import org.apache.spark.streaming.api.java.JavaPairDStream;
import org.apache.spark.streaming.api.java.JavaPairReceiverInputDStream;
import org.apache.spark.streaming.api.java.JavaStreamingContext;
import org.apache.spark.streaming.kafka.KafkaUtils;

import scala.Tuple2;

import com.produban.akamai.api.Alert;
import com.produban.akamai.entity.Akamai;
import com.produban.api.data.Rule;
import com.produban.api.general.K;
import com.produban.indexer.api.Indexer;
import com.produban.indexer.elastic.ElasticIndexer;
import com.produban.indexer.mongo.MongoIndexer;
import com.produban.util.JsonUtil;

public class AkamaiKafkaStreaming implements Serializable {
	HashMap<String, String> configurations;
	ArrayList<Rule> rules;
	ArrayList<Indexer> indexers;

	public AkamaiKafkaStreaming(HashMap<String, String> configurations,
			ArrayList<Rule> rules) {
		super();
		this.configurations = configurations;
		this.rules = rules;
		initIndexers();
	}
	
	private void initIndexers() {
		indexers = new ArrayList<>();
		String typeIndexer = configurations.get(K.SYSTEM.PROPERTY_INDEXER_TYPE);
		if (typeIndexer.contains("elastic")){
			indexers.add(initElasticSearch());
		}
		
		if (typeIndexer.contains("mongo")){
			indexers.add(initMongo());
		}		
	}

	private Indexer initElasticSearch(){
		Indexer indexer;
		String clusterName = configurations.get(K.ELASTIC_SEARCH.PROPERTY_INDEXER_CLUSTER_NAME);
		String indexName = configurations.get(K.ELASTIC_SEARCH.PROPERTY_INDEXER_INDEX_NAME);
		String elasticNodes = configurations.get(K.ELASTIC_SEARCH.PROPERTY_INDEXER_NODES);
		indexer = new ElasticIndexer(clusterName, indexName, "logs", elasticNodes);
		return indexer;
	}
	
	private Indexer initMongo(){
		Indexer indexer;
		String clusterName = configurations.get(K.MONGO.PROPERTY_INDEXER_MONGO_CLUSTER_NAME);
		String mongoDb = configurations.get(K.MONGO.PROPERTY_INDEXER_MONGO_DB);
		String collectionMongo = configurations.get(K.MONGO.PROPERTY_INDEXER_MONGO_COLLECTION);
		indexer = new MongoIndexer(clusterName, mongoDb, collectionMongo);
		return indexer;
	}

	public AkamaiKafkaStreaming() {
		super();
	}

	public void startStreaming() {
		Duration batchDuration = new Duration(5000);
		SparkConf conf = new SparkConf().setMaster("local[2]")
				.setAppName("App");
		JavaStreamingContext ssc = new JavaStreamingContext(conf, batchDuration);

		JavaPairReceiverInputDStream<String, String> pair = KafkaUtils
				.createStream(ssc, getZookeeperNodes(), "myGroup",
						getKafkaTopics());
		executeRules(pair);

		ssc.start();
		ssc.awaitTermination();
	}

	private void executeRules(
			final JavaPairReceiverInputDStream<String, String> streaming) {

		for (final Rule rule : rules) {
			JavaDStream<String> bodyKafka = streaming
					.map(new Function<Tuple2<String, String>, String>() {
						@Override
						public String call(Tuple2<String, String> v1)
								throws Exception {
							return v1._2;
						}
					});

			// bodyKafka.print();
			JavaDStream<String> filteredLines = bodyKafka
					.filter(new Function<String, Boolean>() {
						@Override
						public Boolean call(String v1) throws Exception {
							return v1.contains(rule.getRegex());
						}
					});

			JavaDStream<Akamai> filteredAkamai = filteredLines
					.map(new Function<String, Akamai>() {
						@Override
						public Akamai call(String v1) throws Exception {
							return JsonUtil.read(v1.getBytes(), Akamai.class);
						}
					});

			// Parse to a list of tuples<String(ip_url), AkamaiObj>
			JavaPairDStream<String, Akamai> tuplesAkamai = filteredAkamai
					.mapToPair(new PairFunction<Akamai, String, Akamai>() {
						@Override
						public Tuple2<String, Akamai> call(Akamai jsonAkamai)
								throws Exception {

							String srcIp = jsonAkamai.getMessage().getCliIP();
							String srcURL = jsonAkamai.getMessage()
									.getReqHost();
							Tuple2<String, Akamai> tuple = new Tuple2<String, Akamai>(
									srcIp + "_" + srcURL, jsonAkamai);
							return tuple;
						}
					});

			Duration window = new Duration(rule.getWindow());
			Duration slideDuration = new Duration(rule.getSlideWindow());
			JavaPairDStream<String, Iterable<Akamai>> tupleStreaming = tuplesAkamai
					.groupByKeyAndWindow(window, slideDuration);

			tupleStreaming
					.foreachRDD(new Function<JavaPairRDD<String, Iterable<Akamai>>, Void>() {
						@Override
						public Void call(
								JavaPairRDD<String, Iterable<Akamai>> rdd)
								throws Exception {

							List<Tuple2<String, Iterable<Akamai>>> elem = rdd
									.take(1);

							if (elem.size() > 0) {
								System.out.println("XXX elem " + elem.size());
								List<Akamai> elements = IteratorUtils.toList(elem
										.get(0)._2.iterator());
								System.out.println("XXX elements " + elements.size());
								if (rule.getNumberTimes() <= elements.size()) {
									System.out.println("XXX insertando.. " + elements.size());
									createAlert(rule.getMessage(),elements, indexers);
								}
							}

							return null;
						}
					});
		}
	}

	private void createAlert(final String message,
			final List<Akamai> jsonsDetected, final List<Indexer> Indexer) {
		Akamai[] akamayArray = new Akamai[jsonsDetected.size()];
		jsonsDetected.toArray(akamayArray);		
		Alert alert = new Alert(message, akamayArray);
		String jsonAlert = JsonUtil.write(alert);
		
		for (Indexer indexer : indexers){
			indexer.indexJson(jsonAlert);
		}
	}


	public Map<String, String> getConfigurations() {
		return configurations;
	}

	public List<Rule> getRules() {
		return rules;
	}

	public String getZookeeperNodes() {
		return configurations.get(K.KAFKA.PROPERTY_ZOOKEEPER_NODES);
	}

	public Map<String, Integer> getKafkaTopics() {
		Map<String, Integer> map = new HashMap<>();
		map.put("flume-topic", 1);
		return map;
	}
}
